module/VuFind/src/VuFind/ILS/Driver/Folio.php

```diff
@@ -542,6 +542,43 @@ protected function fetchUserWithCql($query)
         return count($json->users) === 1 ? $json->users[0] : null;
     }
 
+    /**
+     * Helper function to retrieve paged results from FOLIO API
+     *
+     * @param string $responseKey Key containing values to collect in response
+     * @param string $interface   FOLIO api interface to call
+     * @param array  $query       CQL query
+     *
+     * @return array
+     */
+    protected function getPagedResults($responseKey, $interface, $query = [])
+    {
+        $count = 0;
+        $limit = 1000;
+        $offset = 0;
+
+        do {
+            $combinedQuery = array_merge($query, compact('offset', 'limit'));
+            $response = $this->makeRequest(
+                'GET',
+                $interface,
+                $combinedQuery
+            );
+            $json = json_decode($response->getBody());
+            if ($json == null) {
+                throw new ILSException('Could parse response from ILS');
+            }
+            $total = $json->totalRecords ?? 0;
+            foreach ($json->$responseKey ?? [] as $item) {
+                $count++;
+                if ($count % $limit == 0) {
+                    $offset += $limit;
+                }
+                yield $item ?? '';
+            }
+        } while ($count < $total);
```

Comment:
One other thought: maybe we should add a safety check that bails out of this loop if we iterate a certain number of times without adding anything new... otherwise, a malformed response or incorrect $responseKey value could lead to an infinite loop.

Reply:
Thank you. I was a little slow to come around to understanding the logic in the function I'm trying to copy, I think I've more or less replicated what you've done for the reserves loops now. Let me know what you think.